[motion-path] Refactor various parts of path operation and RenderStyle

nmoucht opened the pull request:
#### 3f5059d542cf81542755719fab60d75462d24768
<pre>
[motion-path] Refactor various parts of path operation and RenderStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=260818">https://bugs.webkit.org/show_bug.cgi?id=260818</a>
rdar://114584136

Reviewed by Simon Fraser.

This refactoring addresses a few issues with the current motion path design. First,
for paths that need additional data including its containing block rect or the element&apos;s
offset from its containing block, it is not correct to use the path operation to hold
these values, as the path operation is supposed to represent just the path specified by the
css. Instead, pass down these necessary values through a TransformOperationData struct.
Also added a test for this case to ensure that style sharing is not broken after this change.
Also, factor out some of the code related to computing the ray path into the new MotionPath
class, which now also includes applyMotionPathTransform.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-019-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-019.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::computeTransformedExtentViaMatrix const):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::computedTransform):
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::AcceleratedEffectValues):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::RayPathOperation::create):
(WebCore::RayPathOperation::clone const):
(WebCore::RayPathOperation::getPath const):
(WebCore::RayPathOperation::lengthForPath const): Deleted.
(WebCore::RayPathOperation::lengthForContainPath const): Deleted.
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::getPath): Deleted.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::applyTransform const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateTransform):
(WebCore::RenderLayer::setReferenceBoxForPathOperations): Deleted.
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::applySVGTransform const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::applyTransform const):
(WebCore::RenderStyle::applyCSSTransform const):
(WebCore::getTraversalStateAtDistance): Deleted.
(WebCore::RenderStyle::applyMotionPathTransform const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::animatedLocalTransform const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/267416@main">https://commits.webkit.org/267416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b8d65c212958bcf9ed4bc1b1ffbbb48d9435e69

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16519 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16841 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17281 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18300 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15493 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16709 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20072 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16984 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17839 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16716 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17126 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14285 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19078 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14370 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14969 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15358 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15135 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19450 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15733 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13357 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14933 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14809 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3956 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19302 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15558 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -0,0 +1,161 @@
+            data.containingBlockBoundingRect = pathReferenceBoxRect;
+            auto left = renderer.style().left();
+            auto top = renderer.style().top();
+            data.rayStartingPosition = FloatPoint(left.isPercent() ? left.value() / 100 * pathReferenceBoxRect.width() : left.value(), top.isPercent() ? top.value() / 100 * pathReferenceBoxRect.height() : top.value());
```
Can you use `floatValueForLength()` here?

nt1m replied:
You could just use `floatPointForLengthPoint` like I did in #16953

nmoucht replied:
Right this will get updated after https://github.com/WebKit/WebKit/pull/16953 lands.

smfr commented on Source/WebCore/rendering/MotionPath.h:
```diff
@@ -0,0 +1,73 @@
+public:
+    static std::optional<MotionPathData> populateMotionPathDataForRenderer(const RenderElement&);
+    static bool needsUpdateAfterContainingBlockLayout(PathOperation*);
+    static void applyMotionPathTransform(const RenderStyle&, TransformationMatrix&, const TransformOperationData&);
```
Put the `TransformationMatrix` argument last.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/MotionPath.h:
```diff
@@ -0,0 +1,73 @@
+class MotionPath {
+public:
+    static std::optional<MotionPathData> populateMotionPathDataForRenderer(const RenderElement&);
+    static bool needsUpdateAfterContainingBlockLayout(PathOperation*);
```
`const PathOperation*`. Could this take a const reference?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -0,0 +1,161 @@
+    auto& boundingBox = data.containingBlockBoundingRect;
+    auto distances = distanceOfPointToSidesOfRect(boundingBox, data.rayStartingPosition);
+
+    switch (rayPathOperation.size()) {
```
Weird that something called `size` doesn't return a size type (like `FloatSize`).

nmoucht replied:
This is defined by the spec as ray-size so I think its a reasonable name.

smfr commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -0,0 +1,161 @@
+    case RayPathOperation::Size::FarthestSide:
+        return std::max( { distances.top(), distances.bottom(), distances.left(), distances.right() } );
+    case RayPathOperation::Size::FarthestCorner:
+        return std::sqrt(std::pow(std::max(distances.left(), distances.right()), 2) + std::pow(std::max(distances.top(), distances.bottom()), 2));
```
This should use functions from `GeometryUtilities.h`

nmoucht replied:
Fixed to use std::hypot.

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -0,0 +1,161 @@
+    return std::nullopt;
+}
+
+static PathTraversalState getTraversalStateAtDistance(const Path& path, const Length& distance)
```
Can we remove the `get` prefix while we're here? This is against the webkit style.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/MotionPath.h:
```diff
@@ -0,0 +1,73 @@
+
+struct MotionPathData {
+    FloatRect containingBlockBoundingRect;
+    FloatPoint rayStartingPosition;
```
Should this be specific to `ray` ? the concept of "offset starting position" exists for (almost) all shapes, would this be re-used? https://drafts.fxtf.org/motion-1/#offset-starting-position

nmoucht replied:
Yeah changed to use offsetFromContainingBlock.

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -0,0 +1,161 @@
+
+namespace WebCore {
+
+std::optional<MotionPathData> MotionPath::populateMotionPathDataForRenderer(const RenderElement& renderer)
```
I would just name this `motionPathDataForRenderer`

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/MotionPath.h:
```diff
@@ -0,0 +1,79 @@
+        : m_boundingBox(boundingBox)
+    {
+        if (renderer)
+            m_containingBlockData = MotionPath::populateMotionPathDataForRenderer(*renderer);
```
Feels a bit odd this named m_containingBlockData when the struct is named MotionPathData

nmoucht replied:
Changed to m_motionPathData.

nt1m commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -236,13 +231,7 @@ class RayPathOperation final : public PathOperation {
-
-    WEBCORE_EXPORT static Ref<RayPathOperation> create(float angle, Size, bool isContaining, FloatRect&& containingBlockBoundingRect, FloatPoint&& position);
-
+    WEBCORE_EXPORT static Ref<RayPathOperation> create(float angle, Size size, bool isContaining);
```
Let's remove the `size` name as the style checker says :)

nmoucht replied:
Fixed.

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -5225,6 +5225,7 @@ webkit.org/b/233340 imported/w3c/web-platform-tests/css/motion/offset-anchor-tra
 
 # CSS motion path ray test failing due to getting wrong size from containing block.
 webkit.org/b/233344 imported/w3c/web-platform-tests/css/motion/offset-path-ray-007.html [ ImageOnlyFailure ]
+webkit.org/b/233344 imported/w3c/web-platform-tests/css/motion/offset-path-ray-009.html [ ImageOnlyFailure ]
```
Do we know why this regression is happening?

nmoucht replied:
I think this is because of removing the hack in setReferenceBoxForPathOperations. Regardless this will be fixed after https://github.com/WebKit/WebKit/pull/16633 lands.

nmoucht committed:
[motion-path] Refactor various parts of path operation and RenderStyle
https://bugs.webkit.org/show_bug.cgi?id=260818
rdar://114584136

Reviewed by Simon Fraser.

This refactoring addresses a few issues with the current motion path design. First,
for paths that need additional data including its containing block rect or the element's
offset from its containing block, it is not correct to use the path operation to hold
these values, as the path operation is supposed to represent just the path specified by the
css. Instead, pass down these necessary values through a TransformOperationData struct.
Also added a test for this case to ensure that style sharing is not broken after this change.
Also, factor out some of the code related to computing the ray path into the new MotionPath
class, which now also includes applyMotionPathTransform.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-019-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-019.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::computeTransformedExtentViaMatrix const):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::computedTransform):
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::AcceleratedEffectValues):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::RayPathOperation::create):
(WebCore::RayPathOperation::clone const):
(WebCore::RayPathOperation::getPath const):
(WebCore::RayPathOperation::lengthForPath const): Deleted.
(WebCore::RayPathOperation::lengthForContainPath const): Deleted.
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::getPath): Deleted.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::applyTransform const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateTransform):
(WebCore::RenderLayer::setReferenceBoxForPathOperations): Deleted.
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::applySVGTransform const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::applyTransform const):
(WebCore::RenderStyle::applyCSSTransform const):
(WebCore::getTraversalStateAtDistance): Deleted.
(WebCore::RenderStyle::applyMotionPathTransform const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::animatedLocalTransform const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/267416@main
